### PR TITLE
RFC56: Add support for custom driver annotations for CNA data

### PR DESCRIPTION
**Pull request by pvannierop**
# Add support for custom driver annotations for CNA data

Note:
- This PR is part of RFC56
- This PR covers the following sub-task in RFC56:
   1. [Custom PD annotations for CNA alteration type](https://docs.google.com/document/d/1TLRIOurdRK1HE7nDPqk6d4zhfjpgchBd9tU6PQpbwxw/edit#heading=h.z9ypreqxi1s0)

This PR will:
- Update the database schema to hold custom driver annotations for mutations, discrete copy number alterations and structural variants to the _alteration_driver_annotation_
image
- Add support in validator for custom driver annotations for discrete CNA data
- Update import code for for custom driver annotations for discrete CNA data
- Update MyBatis mappers to reflect schema update
- Update API for discrete CNA data to return custom driver annotations

### CODE REVIEW

**Review by adamabeshouse — APPROVED** [review]

**Comment by pvannierop** [comment]
@adamabeshouse @inodb @alisman I will address the one open question today (the one about the remaining FIXME). After that this PR can be merged. I will poke you when I am done.

**Comment by jjgao** [comment]
Before merging, we should discuss our goal of generalizing (https://github.com/cBioPortal/cbioportal/issues/7897) and define the path towards the goal.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCnaEvent.java [hunk 1]
```diff
@@ -147,38 +159,53 @@ private static long addCnaEventDirectly(CnaEvent cnaEvent) throws DaoException {
...
+                    + " sample_cna_event.SAMPLE_ID,"
+                    + " sample_cna_event.GENETIC_PROFILE_ID,"
+                    + " ENTREZ_GENE_ID,"
+                    + " ALTERATION,"
+                    + " alteration_driver_annotation.DRIVER_FILTER,"
+                    + " alteration_driver_annotation.DRIVER_FILTER_ANNOTATION,"
+                    + " alteration_driver_annotation.DRIVER_TIERS_FILTER,"
+                    + " alteration_driver_annotation.DRIVER_TIERS_FILTER_ANNOTATION"
+                    + " FROM sample_cna_event"
+                    + " LEFT JOIN alteration_driver_annotation ON"
```
Comment (averyniceday) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCnaEvent.java:172]: has this query been tested on a db with large number of records/large study? Just wondering if there was any changes in performance that needs to be considered for periodic imports.

**Review by averyniceday — APPROVED** [review]
Looks good to me :) Just a small question if the new query has been tested for performance.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCnaEvent.java (hunk 1 above)
Comment (pvannierop) [reply core/src/main/java/org/mskcc/cbio/portal/dao/DaoCnaEvent.java:172]: No, it has not. Is there a particular way to test this beforehand that you can recommend?

**Comment by inodb** [comment]
Hey @pvannierop - this seems good to go. Nice work! We'll have to schedule the db migration for deployment but code looks good. Did you want to make any more changes or can we merge it?

**Comment by inodb** [comment]
There are a few screenshots failing in the frontend e2e tests related to this PR, we want to make sure those changes are expected

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCnaEvent.java (hunk 1 above)
Comment (inodb) [reply core/src/main/java/org/mskcc/cbio/portal/dao/DaoCnaEvent.java:172]: btw i did notice this stackoverflow issue re doing a left join with empty rows: https://stackoverflow.com/questions/31207935/mysql-with-left-join-is-slow-with-empty-row. Didn't read carefully to check if this applies here but just sharing for reference. I guess in the public cBioPortal `alteration_driver_annotation` will be empty at first

Could also try with e.g. the latest dump of the public portal: http://download.cbioportal.org/mysql-snapshots/mysql-snapshots-toc.html

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoMutation.java [hunk 2]
```diff
@@ -1548,7 +1556,7 @@ public static void deleteAllRecords() throws DaoException {
                 try {
                     con = JdbcUtil.getDbConnection(DaoMutation.class);
                     pstmt = con.prepareStatement(
-                            "SELECT DISTINCT DRIVER_TIERS_FILTER FROM mutation "
+                            "SELECT DISTINCT DRIVER_TIERS_FILTER FROM alteration_driver_annotation "
```
Comment (alisman) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoMutation.java:1559]: @pvannierop @dippindots this DAO only has to do with import right?  This is not related to reading db, right?

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml [hunk 3]
```diff
@@ -61,6 +61,10 @@
         INNER JOIN sample ON mutation.SAMPLE_ID = sample.INTERNAL_ID
         INNER JOIN patient ON sample.PATIENT_ID = patient.INTERNAL_ID
         INNER JOIN cancer_study ON patient.CANCER_STUDY_ID = cancer_study.CANCER_STUDY_ID
+        LEFT JOIN alteration_driver_annotation ON
```
Comment (alisman) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml:64]: @pvannierop is it possible that we haven't introduced appropriate indexing on this table (alteration_driver_annotation) ?
